mm2src/mm2_metrics/src/recorder.rs

```diff
@@ -0,0 +1,211 @@
+            new_lib::{MetricType, MetricsJson}};
+use common::log::Tag;
+use fomat_macros::wite;
+use itertools::Itertools;
+use metrics::{Counter, Gauge, Key, KeyName, Label, Recorder, Unit};
+use metrics_core::ScopedString;
+use metrics_exporter_prometheus::formatting::key_to_parts;
+use metrics_util::registry::{GenerationalAtomicStorage, Registry};
+use std::fmt::Write;
+
+pub struct Snapshot {
+    pub counters: HashMap<String, HashMap<Vec<Label>, u64>>,
+    pub gauges: HashMap<String, HashMap<Vec<Label>, u64>>,
+    pub histograms: HashMap<String, HashMap<Vec<Label>, f64>>,
+}
+
+pub struct MmRecorder {
+    pub(crate) registry: Registry<Key, GenerationalAtomicStorage>,
+}
+
+impl Default for MmRecorder {
+    fn default() -> Self {
+        Self {
+            registry: Registry::new(metrics_util::registry::GenerationalStorage::atomic()),
+        }
+    }
+}
+
+impl MmRecorder {
+    pub fn get_metrics(&self) -> Snapshot {
+        let counters = self
+            .registry
+            .get_counter_handles()
+            .into_iter()
+            .map(|(key, counter)| {
+                counter.get_generation();
+                let value = counter.get_inner().load(Ordering::Acquire); // This is a specific part for counter/gauge and histogram.
+                let inner = key_value_to_snapshot_entry(key.clone(), value);
+                (key.to_string(), inner)
+            })
+            .collect::<HashMap<String, HashMap<_, _>>>();
+
+        let gauges = self
+            .registry
+            .get_gauge_handles()
+            .into_iter()
+            .map(|(key, gauge)| {
+                gauge.get_generation();
+                let value = gauge.get_inner().load(Ordering::Acquire); // This is a specific part for counter/gauge and histogram.
+                let inner = key_value_to_snapshot_entry(key.clone(), value);
+                (key.to_string(), inner)
+            })
+            .collect::<HashMap<String, HashMap<_, _>>>();
+
+        let histograms = self
+            .registry
+            .get_histogram_handles()
+            .into_iter()
+            .map(|(key, histogram)| {
+                histogram.get_generation();
+                let value: f64 = histogram.get_inner().data().iter().sum(); // This is a specific part for counter/gauge and histogram.
+                let inner = key_value_to_snapshot_entry(key.clone(), value);
+                (key.to_string(), inner)
+            })
+            .collect::<HashMap<String, HashMap<_, _>>>();
+
+        Snapshot {
+            counters,
+            gauges,
+            histograms,
+        }
+    }
+
+    pub fn prepare_tag_metrics(&self) -> Vec<PreparedMetric> {
+        let mut output: Vec<PreparedMetric> = vec![];
+
+        for (key, counter) in self.registry.get_counter_handles() {
+            let value = counter.get_inner().load(Ordering::Acquire);
+            output.push(map_metric_to_prepare_metric(key, value));
+        }
```

Comment:
Please take a look at the previous `TagObserver` structure:
https://github.com/KomodoPlatform/atomicDEX-API/blob/b81acc6a54cae7a31e88dffb8430d37dcf471d33/mm2src/common/mm_metrics/native.rs#L198-L205
Here `metrics` is a map where each `MetricNameValueMap` (e.g. `HashMap<MetricName, Integer>`) is mapped by `MetricLabels`.

For example, we have the following metrics:
> Metric.name=rpc_client.request.count, Metric.value=7649, Metric.labels=[coin=tBCH client=electrum]
Metric.name=rpc_client.response.count, Metric.value=7615, Metric.labels=[coin=tBCH client=electrum]
Metric.name=rpc_client.traffic.in, Metric.value=335206, Metric.labels=[coin=tBCH client=electrum]
Metric.name=rpc_client.traffic.out, Metric.value=700604, Metric.labels=[coin=tBCH client=electrum]
Metric.name=rpc_client.request.count, Metric.value=6345, Metric.labels=[coin=tQTUM client=electrum]
Metric.name=rpc_client.response.count, Metric.value=6935, Metric.labels=[coin=tQTUM client=electrum]
Metric.name=rpc_client.traffic.in, Metric.value=2042993, Metric.labels=[coin=tQTUM client=electrum]
Metric.name=rpc_client.traffic.out, Metric.value=633840, Metric.labels=[coin=tQTUM client=electrum]

To print them to the log compactly, we decided to combine them by labels.
So here is the example of the log output:
> · 2022-06-21 18:27:44 +0200 [coin=tBCH client=electrum] rpc_client.request.count=7649 rpc_client.response.count=7615 rpc_client.traffic.in=335206 rpc_client.traffic.out=700604
· 2022-06-21 18:27:44 +0200 [coin=tQTUM client=electrum] rpc_client.request.count=6345 rpc_client.response.count=6935 rpc_client.traffic.in=2042993 rpc_client.traffic.out=633840

Reply:
I've prepared an example how `TagObserver` should be implemented:
```rust
struct TagObserver;

impl TagObserver {
    fn log_tag_metrics(log_state: &LogArc, recorder: &MmRecorder) {
        Self::prepare_tag_metrics(recorder)
            .into_iter()
            .for_each(|(labels, name_value_map)| {
                let tags = labels_to_tags(labels.iter());
                let message = name_value_map_to_message(name_value_map);

                log_state.log_deref_tags("", tags, &message);
            })
    }

    fn prepare_tag_metrics(recorder: &MmRecorder) -> HashMap<MetricLabels, MetricNameValueMap> {
        let mut output = HashMap::new();

        for (key, counter) in recorder.registry.get_counter_handles() {
            // The following implementation can be wrapped into a generic function (even for histograms).
            let value = counter.get_inner().load(Ordering::Acquire);

            let (metric_name, labels) = key.into_parts();
            output
                .entry(labels)
                .or_insert_with(MetricNameValueMap::new)
                .insert(metric_name.as_str().to_owned(), PreparedMetric::Metric(value));
        }

        // Do the same for gauges and histograms.
        // Please note that I made a mistake in the previous histogram tag observing,
        // because they should be mapped into `HashMap<MetricLabels, MetricNameValueMap>` instead of `HashMap<Key, Histogram<u64>>`.

        output
    }
}
```

Where
```rust
type MetricLabels = Vec<Label>;
type MetricNameValueMap = HashMap<String, PreparedMetric>;
```